### PR TITLE
`strings::join_list_elements` options for empty list inputs

### PR DESCRIPTION
**Pull request by ttnghia**
This PR implements a new option for `strings::join_list_elements` on top of https://github.com/rapidsai/cudf/pull/8282. In particular, the new option is:
```
/**
 * @brief Setting for specifying what will be output from `join_list_elements` when an input list
 * is empty.
 */
enum class output_if_empty_list {
  EMPTY_STRING,  ///< Empty list will result in empty string
  NULL_ELEMENT   ///< Empty list will result in a null
};
```
This new option is necessary for implementing `concat_ws` in Spark, since the behavior of the output string is required to be different depending on the situation.

Currently blocked from merging by https://github.com/rapidsai/cudf/pull/8282.

### CODE REVIEW

**Review by robertmaynard — APPROVED** [review]
Approved for cudf-cmake-codeowners

**Review by kkraus14 — APPROVED** [review]
Cython changes LGTM, thanks @ttnghia

**Comment by hyperbolic2346** [comment]
This is marked a non-breaking change my, but isn't the change to concatenate to no longer allow a single column breaking?

**Comment by ttnghia** [comment]
> This is marked a non-breaking change my, but isn't the change to concatenate to no longer allow a single column breaking?

Sorry for the confusion. That breaking change is in the PR https://github.com/rapidsai/cudf/pull/8282. When it is merged, the difference in this one will be less, which will only include changes for adding the new option in `join_list_elements`.

**Comment by ttnghia** [comment]
Rerun tests.

**Review by hyperbolic2346 — CHANGES_REQUESTED** [review]
I don't see any tests for `output_if_empty_list::EMPTY_STRING`. Am I missing them or are they excluded for a reason?

**Comment by ttnghia** [comment]
> I don't see any tests for `output_if_empty_list::EMPTY_STRING`. Am I missing them or are they excluded for a reason?

It is the default behavior, so it is tested if there is no specifying `...::NULL_ELEMENT`.

**Review by hyperbolic2346 — APPROVED** [review]

**Review by davidwendt — APPROVED** [review]

**Comment by ttnghia** [comment]
Rerun tests.

**Comment by ttnghia** [comment]
@gpucibot merge

**Comment by ttnghia** [comment]
Rerun tests.

**Review by kkraus14 — CHANGES_REQUESTED** [review]
Since the behavior changed would be good to confirm this is expected behavior from the Python side.

@shwina @galipremsagar could one of you take a look and share thoughts regarding the behavior?

File: python/cudf/cudf/core/column/string.py [hunk 1]
```diff
@@ -485,10 +485,11 @@ def join(
...
         string_na_rep : str, default None
-            This character will take the place of any null strings
-            (not empty strings) in the Series.
-            If ``string_na_rep`` is ``None``, it defaults to empty
-            space "".
+            This character will take the place of null strings
+            (not empty strings) in the Series but will be considered
+            only if the Series contains at least one non-null string.
+            If ``string_na_rep`` is ``None`` or the Series contains only
+            ``None``, it defaults to empty space "".
```
Comment (shwina) [inline python/cudf/cudf/core/column/string.py:492]: It would be really helpful to document this behaviour in an example.

Comment (ttnghia) [reply python/cudf/cudf/core/column/string.py:492]: Thanks @shwina. There are already several examples demonstrating such behavior---null replacement if there are both Null and valid strings, and empty return string (ignore null replacement) if there are only Null.

Comment (shwina) [reply python/cudf/cudf/core/column/string.py:492]: I meant in the [`Examples`](https://github.com/rapidsai/cudf/blob/07b8c9c646f2f15acb0821b42feda462880bd745/python/cudf/cudf/core/column/string.py#L514) section of the docstring for this function. I don't think we have an example for the "all nulls" case described.

Comment (ttnghia) [reply python/cudf/cudf/core/column/string.py:492]: Sure. Done.

**Review by galipremsagar — APPROVED** [review]

**Comment by ttnghia** [comment]
> Is there any way to preserve the original behaviour from the Python side? Perhaps make the original behaviour opt-in in libcudf?

The C++ API allows to do that. I will need to change the python interface (it was setup to call C++ API with default parameters, which returns an empty string on all-null lists).

**Review by shwina — APPROVED** [review]
LGTM!

**Review by kkraus14 — APPROVED** [review]

**Comment by kkraus14** [comment]
@gpucibot merge

**Comment by ttnghia** [comment]
Rerun tests.

**Comment by ttnghia** [comment]
Rerun tests.

**Comment by ttnghia** [comment]
Rerun tests.

**Comment by ttnghia** [comment]
Rerun tests.

**Comment by ttnghia** [comment]
Rerun tests.

**Comment by ttnghia** [comment]
Rerun tests.

**Comment by ttnghia** [comment]
Rerun tests.

**Comment by kkraus14** [comment]
rerun tests

**Comment by kkraus14** [comment]
@gpucibot merge

**Pull request merged**